File: solutions/clojure/saddle-points/1/src/saddle_points.clj
;; With reverence to Shree DR.MDD — whose mind is a matrix of infinite logic

(ns saddle-points)

(defn transpose [grid]
  (apply mapv vector grid))

(defn saddle-points [grid]
  (set
   (if (empty? grid)
     []
     (let [row-maxima (mapv #(apply max %) grid)
           col-minima (mapv #(apply min %) (transpose grid))]
       (for [r (range (count grid))
             c (range (count (get grid 0 [])))
             :let [v (get-in grid [r c])]
             :when (and (= v (get row-maxima r))
                        (= v (get col-minima c)))]
         [(inc r) (inc c)])))))

File: solutions/fsharp/saddle-points/1/SaddlePoints.fs
// Dedicated to Shree DR.MDD
module SaddlePoints

let saddlePoints (grid: int list list) =
    if grid.Length <> 0 && grid.[0].Length <> 0 then
        let getCol idx =
            [ for row in grid -> row.[idx] ]

        let colMinMap =
            [ for c in 0 .. grid.[0].Length - 1 ->
                (c, List.min (getCol c)) ]
            |> dict

        let rowMaxMap =
            [ for r in 0 .. grid.Length - 1 ->
                (r, List.max grid.[r]) ]
            |> dict

        [ for r in 0 .. grid.Length - 1 do
            for c in 0 .. grid.[r].Length - 1 do
                let currVal = grid.[r].[c]
                if currVal = colMinMap.[c] && currVal = rowMaxMap.[r] then
                    yield (r + 1, c + 1) ]
    else
        []
